misc/python/materialize/mzexplore/clone.py

```diff
@@ -0,0 +1,155 @@
+    mzfmt: bool,
+) -> None:
+    """
+    Generate a DDL script that clones object_ids and dependencies that live in
+    the same cluster into a different cluster.
+
+    The catalog items included in the script will be seeded from the .
+    """
+
+    database_new = sql.identifier(database, True)
+    schema_new = sql.identifier(schema, True)
+    cluster_new = sql.identifier(cluster, True)
+
+    with closing(
+        sql.Database(
+            port=db_port,
+            host=db_host,
+            user=db_user,
+            database=None,
+            password=db_pass,
+            require_ssl=db_require_ssl,
+        )
+    ) as db:
+        output_template = string.Template(
+            textwrap.dedent(
+                """
+                -- original id: $id
+                $create_sql
+                """
+            ).lstrip()
+        )
+
+        # Extract dependencies
+        # --------------------
+
+        if len(object_ids) == 0:
+            msg = "At least one object_id necessary"
+            raise ValueError(msg)
+
+        old_clusters = list(db.object_clusters(object_ids))
+        if len(old_clusters) != 1:
+            msg = f"Cannot find unique old cluster for object ids: {object_ids}"
+            raise ValueError(msg)
+
+        [cluster_old] = old_clusters
+
+        if cluster_old["name"] == cluster:
+            msg = f"Old and new clusters can't have the same name: {cluster}"
+            raise ValueError(msg)
+
+        items_sub = dict()
+        for item in db.clone_dependencies(object_ids, cluster_old["id"]):
+            item_database = sql.identifier(item["database"], True)
+            item_schema = sql.identifier(item["schema"], True)
+            item_name = sql.identifier(item["name"], True)
+            fqname = f"{item_database}.{item_schema}.{item_name}"
+
+            try:
+                item_type = ItemType(item["type"])
+            except ValueError:
+                warn(f"Unsupported item type `{item['type']}` for {fqname}")
+                continue
+
+            show_create_query = item_type.show_create(fqname)
+            if show_create_query is None:
+                continue
+
+            if item_type == ItemType.INDEX:
+                assert item_name not in items_sub
+                name_new = sql.identifier(
+                    "_".join(
+                        [
+                            item["name"],
+                            item["id"],
+                        ]
+                    ),
+                    True,
+                )
+                items_sub[item_name] = name_new
+            else:
+                assert fqname not in items_sub
+                name_new = sql.identifier(
+                    "_".join(
+                        [
+                            item["database"],
+                            item["schema"],
+                            item["name"],
+                            item["id"],
+                        ]
```

Comment:
ID's are unique, so database and schema wouldn't be necessary here. Are they included to make the resulting names easier to tell apart?

Reply:
Yes, because at least in the beginning I will be looking at the diff between the original and the modified DDL statemens in order to verify that the substitutions are sound, and it's easier to do that if the diff is between `materialize.public.foo` and `materialize._mz_test.materialize_public_foo_u16` (as opposed to `u16`).